crates/bevy_pbr/src/pbr_material.rs

```diff
@@ -46,6 +46,11 @@ pub struct StandardMaterial {
     #[dependency]
     pub base_color_texture: Option<Handle<Image>>,
 
+    /// The UV channel to use when sampling the base colour texture.
+    ///
+    /// Defaults to 0. Supports 0 or 1.
```

Comment:
Question from a rendering novice:

If this value can be 0 or 1, why are we using an entire u32 for these fields?

Reply:
@alice-i-cecile Excellent question! As far as I can tell, WGSL doesn't have a u8. And I figured it would be a pain to convert from a u8 in Rust land to a u32 in shader land...

HOWEVER, I'm a newb at Rust and WGSL, so I would happily be corrected on this :)

Reply:
Does it have booleans?

In theory we could probably pack these into a more compact representation using bitflags or something as well, but I'm unsure if that would actually be desired.

Reply:
Yes, no doubt you are right.

Personally I would hesitate to use booleans if only because there ought to eventually be support for more than two UV channels.

Reply:
I was thinking about packing several u8s into a single u32 for sending to the shader. But the more I think on it I realise it we have a trade-off between performance and memory usage. Having to pack and unpack the values would add a small performance cost on both CPU and GPU, although it would save some bytes on the GPU side per material.

I suppose we could define these as u8 on StandardMaterial in pbr_material.rs, then cast them to u32 when building the StandardMaterialUniform. I'm not sure what that gets us, but perhaps a little memory saving on Rust side?

Reply:
Mhmm, I wouldn't worry about optimizing this further unless there's actual benchmarks or stress tests showing a difference. The tradeoffs are hard to evaluate!

Reply:
I don't know the perf implications of using separate values vs a packed representation on the uniform side, but knowing that it probably doesn't make sense to support more than 8 uv sets, all of those could be bit-packed with 3 bit each in a single u32, leaving room for additional flags/selectors.

Reply:
I'd love more input on the performance implications. My thinking is that the channel values are changed rarely (probably never after the material is created) but used often (every call to the fragment shader).

I would assume that unpacking adds operations and so wouldn't be ideal for every pixel, every frame. And since there is no u8 in WGSL one would have to unpack a single u32 into multiple u32s regardless, right..?

My knowledge on this level is very limited, so please feel free to correct me!

Something I am very unclear on: Are these values copied from CPU to GPU with each render pass? Or are they kept in GPU memory between passes? (I obviously don't know much about how binding works 🫢)

I hope it's not inappropriate to drag SMEs in here... But um would @IceSentry @superdump or @robtfm care to weigh in on this issue?

Reply:
@robtfm I'd love to see eventual support for more than 2 UV channels, and hypothetically any given texture could have a unique channel value. Do you think this is a feasible thing to aim for in the standard shader or is it simply too much to expect that level of flexibility?

Reply:
Do any other engines support more than 2 uv sets?

Reply:
@robtfm Excellent question! Certainly in the vertex data this is supported by Unity. However, I don't believe the supplied lit shader for URP (the most current "standard" shader) allows you to use those channels for PBR textures. You would instead have to create a custom shader.

Perhaps what I'm trying to achieve here should not be done with the standard material, but with an extended material? I haven't looked into that much, TBH. Would it support such cases?

Reply:
when the actual usage is a uniform switch (i.e. just choosing a uv set for the whole operation, rather than using some logic in the shader), i'm not sure i see the benefit over just setting a different single uv set in your mesh to start with? i guess there's some asset flow / convention that uses this setup?

i think (without contrary evidence from other engines) more than 2 uv sets is probably niche enough that we shouldn't spend extra space / complexity in the standard material to support it. personally i think we should aim to reduce the standard material complexity and make more of the current features into pluggable extensions in future anyway.

for more than 2 sets, it could be done with a custom material or an ExtendedMaterial - the custom_vertex_attribute and extended_material examples are probably the best place to start for that.

Reply:
@robtfm makes sense.

FWIW: My initial motive behind this PR was that I have a set of assets which used two UV sets: a tiled set and a full-unwrap set. The tiled sets are used for most textures to retain a good detail:texture size ratio. The full-unwrap is used for baked AO as AO maps should not have any overlap. These assets work fine out of the box in three.js, but Bevy draws the AO using the tiling set (UV0), which obviously looks very wrong.

At some point I'll try updating this PR using flags instead of u32s!